### PR TITLE
[GEP-26] Fix cloudprovider secret deployment to not renew WorkloadIdentity token on shoot reconciliation

### PR DESCRIPTION
**Pull request by vpnachev**
**How to categorize this PR?**
/area security
/area ipcei
/label ipcei/workload-identity
/kind bug

**What this PR does / why we need it**:
Fix cloudprovider secret deployment to not renew WorkloadIdentity token on shoot reconciliation.
Currently, all already existing annotations, labels and data fields are removed on each shoot reconciliation, most of these changes triggers the workload identity token requestor controller to renew the token. 

**Which issue(s) this PR fixes**:
Follow up on #10239
Part of #9586 

**Special notes for your reviewer**:
cc @dimityrmirchev 

**Release note**:

```other operator

```

### CODE REVIEW

File: pkg/gardenlet/operation/botanist/secrets.go [hunk 1]
```diff
@@ -495,29 +492,48 @@ func (b *Botanist) DeployCloudProviderSecret(ctx context.Context) error {
...
+					securityv1alpha1constants.AnnotationWorkloadIdentityContextObject: string(shootData),
+				},
+			)
+			secret.Labels = utils.MergeStringMaps(
+				secret.Labels,
+				map[string]string{
+					securityv1alpha1constants.LabelPurpose:                  securityv1alpha1constants.LabelPurposeWorkloadIdentityTokenRequestor,
+					securityv1alpha1constants.LabelWorkloadIdentityProvider: credentials.Spec.TargetSystem.Type,
+				},
+				map[string]string{v1beta1constants.GardenerPurpose: v1beta1constants.SecretNameCloudProvider},
```
Comment (dimityrmirchev) [inline pkg/gardenlet/operation/botanist/secrets.go:510]:
```suggestion
				map[string]string{
					securityv1alpha1constants.LabelPurpose:                  securityv1alpha1constants.LabelPurposeWorkloadIdentityTokenRequestor,
					securityv1alpha1constants.LabelWorkloadIdentityProvider: credentials.Spec.TargetSystem.Type,
					v1beta1constants.GardenerPurpose: v1beta1constants.SecretNameCloudProvider,
				},
```

File: pkg/gardenlet/operation/botanist/secrets.go [hunk 2]
```diff
@@ -495,29 +492,48 @@ func (b *Botanist) DeployCloudProviderSecret(ctx context.Context) error {
...
+			)
+
+			// Preserve the token field if already set, otherwise it is removed
+			// which triggers unwanted renewal of the workload identity token.
+			if token, ok := secret.Data["token"]; ok {
+				data["token"] = token
+			}
+
+			secret.Type = corev1.SecretTypeOpaque
+			secret.Data = utils.MergeStringMaps(secret.Data, data)
```
Comment (dimityrmirchev) [inline pkg/gardenlet/operation/botanist/secrets.go:520]: Why do we merge the old and new fields? Can't we just do the following?

```suggestion
			secret.Data = data
```

**Review by dimityrmirchev — COMMENTED** [review]
Two small comments from my side. I am now wondering if we should provide a utility function that handles the preservation of certain annotations, labels, data fileds w.r.t the tokenrequestor controller because extensions are also going to have to handle the same case. I will try to come with a more generic solution for this one

File: pkg/gardenlet/operation/botanist/secrets.go (hunk 2 above)
Comment (vpnachev) [reply pkg/gardenlet/operation/botanist/secrets.go:520]: Good catch, this is left over from the first version of fix.

**Comment by rfranzke** [comment]
What about the following (which looks a bit cleaner/more straightforward/involves less custom merging)?

```golang
// DeployCloudProviderSecret creates or updates the cloud provider secret in the Shoot namespace
// in the Seed cluster.
func (b *Botanist) DeployCloudProviderSecret(ctx context.Context) error {
	var (
		secret = &corev1.Secret{
			ObjectMeta: metav1.ObjectMeta{
				Name:      v1beta1constants.SecretNameCloudProvider,
				Namespace: b.Shoot.SeedNamespace,
			},
		}
		mutateFunc func()
	)

	switch credentials := b.Shoot.Credentials.(type) {
	case *securityv1alpha1.WorkloadIdentity:
		rawProviderConfig, err := json.Marshal(credentials.Spec.TargetSystem.ProviderConfig)
		if err != nil {
			return fmt.Errorf("failed marshalling WorkloadIdentity provider config: %w", err)
		}

		shootInfo := b.Shoot.GetInfo()
		rawContextObject, err := json.Marshal(securityv1alpha1.ContextObject{
			APIVersion: shootInfo.TypeMeta.GroupVersionKind().GroupVersion().String(),
			Kind:       shootInfo.TypeMeta.Kind,
			Namespace:  ptr.To(shootInfo.ObjectMeta.Namespace),
			Name:       shootInfo.ObjectMeta.Name,
			UID:        shootInfo.ObjectMeta.UID,
		})
		if err != nil {
			return fmt.Errorf("failed marshalling WorkloadIdentity context object: %w", err)
		}

		mutateFunc = func() {
			metav1.SetMetaDataAnnotation(&secret.ObjectMeta, securityv1alpha1constants.AnnotationWorkloadIdentityNamespace, credentials.Namespace)
			metav1.SetMetaDataAnnotation(&secret.ObjectMeta, securityv1alpha1constants.AnnotationWorkloadIdentityName, credentials.Name)
			metav1.SetMetaDataAnnotation(&secret.ObjectMeta, securityv1alpha1constants.AnnotationWorkloadIdentityContextObject, string(rawContextObject))

			metav1.SetMetaDataLabel(&secret.ObjectMeta, securityv1alpha1constants.LabelPurpose, securityv1alpha1constants.LabelPurposeWorkloadIdentityTokenRequestor)
			metav1.SetMetaDataLabel(&secret.ObjectMeta, securityv1alpha1constants.LabelWorkloadIdentityProvider, credentials.Spec.TargetSystem.Type)

			if secret.Data == nil {
				secret.Data = make(map[string][]byte)
			}
			secret.Data["config"] = rawProviderConfig
		}

	case *corev1.Secret:
		mutateFunc = func() {
			secret.Data = credentials.Data
		}

	default:
		return fmt.Errorf("unexpected type %T, should be either Secret or WorkloadIdentity", credentials)
	}

	_, err := controllerutils.GetAndCreateOrMergePatch(ctx, b.SeedClientSet.Client(), secret, func() error {
		metav1.SetMetaDataLabel(&secret.ObjectMeta, v1beta1constants.GardenerPurpose, v1beta1constants.SecretNameCloudProvider)
		secret.Type = corev1.SecretTypeOpaque
		mutateFunc()
		return nil
	})
	return err
}
```

**Comment by vpnachev** [comment]
> What about the following (which looks a bit cleaner/more straightforward/involves less custom merging)?

Looks very nice. Both the PR and the proposal have a disadvantage of not removing some data when switching between workload identity and static credentials (e.g. the workload identity annotations and labels are not cleaned up when migrating between the credentials provider).

@dimityrmirchev have a slightly different proposal how to fix this issue and he will open another PR soon, therefore I am closing this one.

/close

**Pull request closed without merging**